core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -1109,36 +1054,139 @@ private boolean shouldPerformTransactionOperation(StoreStatus storeStatus, Predi
                               });
 
                         AtomicLong writeCount = new AtomicLong();
-                        entryWriteFlowable = entryWriteFlowable.doOnNext(me -> writeCount.incrementAndGet());
-
-                        Flowable<NonBlockingStore.SegmentedPublisher<MarshallableEntry<K, V>>> writeFlowable;
 
-                        if (segmented) {
-                           writeFlowable = entryWriteFlowable
-                                 .groupBy(me -> keyPartitioner.getSegment(me.getKey()))
-                                 .map(SegmentPublisherWrapper::wrap);
-                        } else {
-                           writeFlowable = Flowable.just(SingleSegmentPublisher.singleSegment(entryWriteFlowable));
-                        }
+                        Flowable<NonBlockingStore.SegmentedPublisher<MarshallableEntry<K, V>>> writeFlowable =
+                              createWriteFlowable(entryWriteFlowable, shared, segmented, writeCount, storeStatus);
 
-                        if (trace) {
-                           removeFlowable = removeFlowable.doOnSubscribe(sub ->
-                              log.tracef("Store %s has subscribed to write batch", storeStatus.store));
-                        }
+                        CompletionStage<Void> storeBatchStage = flowableHandler.handleFlowables(storeStatus.store(),
+                              segmentCount(segmented), removeFlowable, writeFlowable);
 
-                        return Single.fromCompletionStage(flowableHandler.handleFlowables(storeStatus.store(),
-                              segmentCount(segmented), removeFlowable, writeFlowable)
-                              .thenApply(ignore2 -> writeCount.get()));
-                     }).first(0L);
+                        return Single.fromCompletionStage(storeBatchStage)
+                              .map(ignore2 -> writeCount.get());
+                        // Only take the last element for the count - ensures all stores are completed
+                     }).last(0L);
 
             },
             this::releaseReadLock
       ).toCompletionStage();
    }
 
-   private <K, V> MVCCEntry<K, V> acquireKeyFromContext(InvocationContext ctx, WriteCommand command, Object key,
-         BiPredicate<WriteCommand, Object> commandKeyPredicate) {
-      if (commandKeyPredicate.test(command, key)) {
+   <K> Flowable<NonBlockingStore.SegmentedPublisher<K>> createRemoveFlowable(Flowable<K> keyRemoveFlowable,
```

Comment:
I didn't say this before, but let's pass the `keyRemoveFlowable` in this method and just pass `flowableToUse` from `batchOperation. That way the logic is in a single method.